data/transactions/verify/txn.go

```diff
@@ -0,0 +1,202 @@
+	"github.com/algorand/go-algorand/util/execpool"
+)
+
+// TxnPool verifies that a SignedTxn has a good signature and that the underlying
+// transaction is properly constructed.
+// Note that this does not check whether a payset is valid against the ledger:
+// a SignedTxn may be well-formed, but a payset might contain an overspend.
+//
+// This version of verify is performing the verification over the provided execution pool.
+func TxnPool(s *transactions.SignedTxn, spec transactions.SpecialAddresses, proto config.ConsensusParams, verificationPool execpool.BacklogPool) error {
+	if err := s.Txn.WellFormed(spec, proto); err != nil {
+		return err
+	}
+
+	zeroAddress := basics.Address{}
+	if s.Txn.Src() == zeroAddress {
+		return errors.New("empty address")
+	}
+
+	if s.Sig != (crypto.Signature{}) && !s.Msig.Blank() {
+		return errors.New("signedtxn should only have one of Sig or Msig")
+	}
+
+	outCh := make(chan error, 1)
+	cx := asyncVerifyContext{s, outCh, &proto}
+	verificationPool.EnqueueBacklog(context.Background(), stxnAsyncVerify, &cx, nil)
+	if err, hasErr := <-outCh; hasErr {
+		return err
+	}
+	return nil
+}
+
+// Txn verifies a SignedTxn as being signed and having no obviously inconsistent data.
+// Block-assembly time checks of LogicSig and accounting rules may still block the txn.
+func Txn(s *transactions.SignedTxn, spec transactions.SpecialAddresses, proto config.ConsensusParams) error {
+	if err := s.Txn.WellFormed(spec, proto); err != nil {
+		return err
+	}
+
+	zeroAddress := basics.Address{}
+	if s.Txn.Src() == zeroAddress {
+		return errors.New("empty address")
+	}
+
+	return stxnVerifyCore(s, &proto)
+}
+
+type asyncVerifyContext struct {
+	s     *transactions.SignedTxn
+	outCh chan error
+	proto *config.ConsensusParams
+}
+
+func stxnAsyncVerify(arg interface{}) interface{} {
+	cx := arg.(*asyncVerifyContext)
+	err := stxnVerifyCore(cx.s, cx.proto)
+	if err != nil {
+		cx.outCh <- err
+	} else {
+		close(cx.outCh)
+	}
+	return nil
+}
+
+func stxnVerifyCore(s *transactions.SignedTxn, proto *config.ConsensusParams) error {
+	numSigs := 0
+	hasSig := false
+	hasMsig := false
+	hasLogicSig := false
+	if s.Sig != (crypto.Signature{}) {
+		numSigs++
+		hasSig = true
+	}
+	if !s.Msig.Blank() {
+		numSigs++
+		hasMsig = true
+	}
+	if !s.Lsig.Blank() {
+		numSigs++
+		hasLogicSig = true
+	}
+	if numSigs == 0 {
+		return errors.New("signedtxn has no sig")
+	}
+	if numSigs > 1 {
+		return errors.New("signedtxn should only have one of Sig or Msig or LogicSig")
+	}
+
+	if hasSig {
+		if crypto.SignatureVerifier(s.Txn.Src()).Verify(s.Txn, s.Sig) {
+			return nil
+		}
+		return errors.New("signature validation failed")
+	}
+	if hasMsig {
+		if ok, _ := crypto.MultisigVerify(s.Txn, crypto.Digest(s.Txn.Src()), s.Msig); ok {
+			return nil
+		}
+		return errors.New("multisig validation failed")
+	}
+	if hasLogicSig {
+		return LogicSig(&s.Lsig, proto, s)
+	}
+	return errors.New("has one mystery sig. WAT?")
```

Comment:
There is no point in this return. It cannot logically happen.
I would suggest restructuring the last two if statements : 
```
if hasMsig {
 		if ok, _ := crypto.MultisigVerify(s.Txn, crypto.Digest(s.Txn.Src()), s.Msig); ok {
 			return nil
 		}
 		return errors.New("multisig validation failed")
 	} else  {
 		return LogicSig(&s.Lsig, proto, s)
 	}
```
The reason is that it would always generate unreachable code, which is undesired.

Reply:
I like the explicitness of the current code, because if we ever add a new signature type, we would instantly notice if we forgot to update this function.